src/raftstore/store/fsm/apply.rs

```diff
@@ -1402,8 +1398,41 @@ impl ApplyDelegate {
 
         let merge = req.get_commit_merge();
         let source_region = merge.get_source();
-        let region_state_key = keys::region_state_key(source_region.get_id());
-        let state: RegionLocalState = match self.engines.kv.get_msg_cf(CF_RAFT, &region_state_key) {
+        let source_region_id = source_region.get_id();
+
+        // No matter whether the source peer has applied to the required index,
+        // it's a race to write apply state in both source delegate and target
```

Comment:
I don't quite understand, can you explain the race in detail?

Reply:
Nothing prevents the source delegate to continue applying logs after required index, which can update the apply index in apply state. So both delegates can update the same delegate at the same time.

Reply:
why both delegates write the same region apply state? why?